[trade area processor] main dev branch for the trade area processor built with Gradle

tdesfont opened the pull request:
The aim of this feature development is to provide a geometry generator around an input center geopoint. Given a Well Known Text column of geopoints POINT(..., ...) in DSS, provide a processor (Prepare recipe step) that will take as input the distance (radius if circular area or width/height if rectangular area) and output the Well Known Text MultiPolygon defining the circular/rectangular area around the input points. User will then be able to visualise the trade area as a standard geometry in DSS and later perform a geo join of points in the created polygons. The build is done in Gradle (inspired from KML plugin) and the processor is in Java (inspired from Numeric Base Conversion plugin). To work inside DSS, one needs to build the jar in dist and then restart DSS.

tdesfont committed:
Setting up gradle and package structure for the java processor

tdesfont committed:
Add gradle files

tdesfont committed:
add processor.js

tdesfont committed:
Add Makefile

tdesfont committed:
Add core logic of the plugin on top of the gradle structure

CircleAreaGenerator.java: Generator class for almost circular shape
GeoUtils.java: Formula computing the endpoint
MyGeoPoint.java: To remove in future versions
RectangularAreaGenerator.java: Generator class for rectangular shape
Added several test classes
Status: To test and recheck, not ready for production

tdesfont committed:
Remove legacy code and iterate on the area processor.
Status: Generating a non parametrized polygon area around the center geopoint column

tdesfont committed:
Added UI parameters for the trade area generator
Status: Generating a parametrized polygon area around the center geopoint column, can choose circular or rectangular area
Next: Need to add the choice on unit (kilometers/miles/meters)

tdesfont committed:
Fixed environment variable DKUINSTALLDIR retrieval

MayeulRousselet commented on Makefile:
```diff
@@ -1,47 +1,9 @@
+PLUGIN_VERSION=0.1.0
+PLUGIN_ID=geospatial-toolkit
 
 plugin:
```
don't forget to add the command to run the unit-tests with gradle

tdesfont replied:
Done in [414d8f64373baf4010cf325f7e71fbc8a9e9be13](url)

tdesfont replied:
Done

MayeulRousselet commented on build.gradle:
```diff
@@ -0,0 +1,116 @@
+sourceSets {
+    main {
+        java {
+            srcDirs = ['src/main/java']
```
Add an intermediary directory (java-processors)

tdesfont replied:
Done

MayeulRousselet commented on js/processor.js:
```diff
@@ -0,0 +1,33 @@
+(function() {
```
What is the use of this file? It still references numerical conversion

MayeulRousselet replied:
Oooooook I found the usage of this file.
Its main purpose is to show the description of the processor in the shaker, as shown here
![Capture d’écran 2021-01-11 à 4 19 39 PM](https://user-images.githubusercontent.com/54799490/104200374-df7f0280-5428-11eb-8b9b-85ba50834724.png)

MayeulRousselet commented on plugin.json:
```diff
@@ -10,8 +10,19 @@
+        {
+            "id": "tradearea",
+            "meta": {
+                "label": "Base conversion",
```
To update

tdesfont replied:
Done

MayeulRousselet commented on src/main/java/com/dataiku/dip/plugins/tradearea/AreaGenerator.java:
```diff
@@ -0,0 +1,75 @@
+    }
+
+    public String generateArea(MyGeoPoint center) {
+        System.out.println("Generating an Rectangular area");
```
Don't log with System.out.println but use a logger as you did in KML, and log variable data (width and height)

tdesfont replied:
Done

MayeulRousselet commented on src/main/java/com/dataiku/dip/plugins/tradearea/AreaGenerator.java:
```diff
@@ -0,0 +1,75 @@
+        MyGeoPoint initGeoPoint = null;
+        MyGeoPoint tmpGeoPoint;
+        StringBuilder str = new StringBuilder();
+        str.append("MULTIPOLYGON((");
```
Why a multipolygon?

tdesfont replied:
Changed to POLYGON

MayeulRousselet commented on src/main/java/com/dataiku/dip/plugins/tradearea/AreaGenerator.java:
```diff
@@ -0,0 +1,75 @@
+        MyGeoPoint tmpGeoPoint;
+        StringBuilder str = new StringBuilder();
+        str.append("MULTIPOLYGON((");
+        double alpha = Math.atan(this.height/this.width);
```
give a better name

tdesfont replied:
Set to diagonalAngle

MayeulRousselet commented on src/main/java/com/dataiku/dip/plugins/tradearea/AreaGenerator.java:
```diff
@@ -0,0 +1,75 @@
+        StringBuilder str = new StringBuilder();
+        str.append("MULTIPOLYGON((");
+        double alpha = Math.atan(this.height/this.width);
+        double[] angles = {90-alpha, 90+alpha, 270-alpha, 270+alpha};
```
Which angles are these?

tdesfont replied:
Detailed in the wiki https://analytics.dataiku.com/projects/RDWIKI/wiki/1092/Plugin:%20GeoSpatial%20Toolkit/edit

MayeulRousselet commented on src/main/java/com/dataiku/dip/plugins/tradearea/AreaGenerator.java:
```diff
@@ -0,0 +1,75 @@
+
+    public String generateArea(MyGeoPoint center) {
+        System.out.println("Generating an Rectangular area");
+        double startLat = center.latitude;
```
bad name, keep center.latitude as it is way clearer

tdesfont replied:
Done

MayeulRousselet commented on src/main/java/com/dataiku/dip/plugins/tradearea/AreaGenerator.java:
```diff
@@ -0,0 +1,75 @@
+    public String generateArea(MyGeoPoint center) {
+        System.out.println("Generating an Rectangular area");
+        double startLat = center.latitude;
+        double startLong = center.longitude;
```
same here

tdesfont replied:
Done

MayeulRousselet commented on src/main/java/com/dataiku/dip/plugins/tradearea/AreaGenerator.java:
```diff
@@ -0,0 +1,75 @@
+        double startLat = center.latitude;
+        double startLong = center.longitude;
+        MyGeoPoint initGeoPoint = null;
+        MyGeoPoint tmpGeoPoint;
```
Don't need to be declared out of the loop

tdesfont replied:
Put it inside

MayeulRousselet commented on src/main/java/com/dataiku/dip/plugins/tradearea/AreaGenerator.java:
```diff
@@ -0,0 +1,75 @@
+        str.append("MULTIPOLYGON((");
+        double alpha = Math.atan(this.height/this.width);
+        double[] angles = {90-alpha, 90+alpha, 270-alpha, 270+alpha};
+        double angle;
```
don't need to declare the variable outta loop

tdesfont replied:
Done

tdesfont committed:
Fixed output format for polygon. Fixed angle conversion for rectangular area. Removed sun.font.decoration.

tdesfont committed:
GeoUtils.java: Expect input distance in km instead of meters
MyGeoPoint.java: Fixed parser for latitude longitude from WKT POINT
TradeAreaProcessor.java: Removed Meters and set up order of arguments

tdesfont committed:
processor.js: Polish the UI naming of the processor as it appears in the recipe panel
TradeAreaProcessor.java: Handle conversion if input are in miles

tdesfont committed:
Added comments and tests

tdesfont committed:
Add tests trigger in Makefile

tdesfont committed:
Create intermediary java-processors

tdesfont committed:
Fix wording in plugin.json

tdesfont committed:
Use dku logger

tdesfont committed:
Use dku logger

tdesfont committed:
Changed computation formula for the radius, should take half of width and height for computation

tdesfont committed:
Added wording and comments

tdesfont committed:
Implement validate

tdesfont committed:
Changed icon, Reuse GeoPoint and GeoPoint.Coords class from dip

tdesfont committed:
Changed icon, minor changes

tdesfont committed:
Added TODO for check null values

tdesfont committed:
Add tests for invalid input

tdesfont committed:
Minor change to test

MayeulRousselet commented on java-processors/src/main/java/com/dataiku/dip/plugins/tradearea/TradeAreaProcessor.java:
```diff
@@ -0,0 +1,199 @@
+                    "# Input column\n" +
+                    "Contains the geopoints on which the trade area is centered\n \n" +
+                    "# Output column\n" +
+                    "Contains created trade areas as WKT polygons (String column)\n\n"+
```
```suggestion
                    "Contains created trade areas in the WKT format\n\n"+
```

MayeulRousselet commented on java-processors/src/main/java/com/dataiku/dip/plugins/tradearea/GeoUtils.java:
```diff
@@ -0,0 +1,43 @@
+
+public class GeoUtils {
+
+    // radians/degrees conversion
```
Field comments must be in the following format:
```suggestion
    /**
     * radians/degrees conversion
     */
```

tdesfont replied:
Ok

MayeulRousselet commented on java-processors/src/main/java/com/dataiku/dip/plugins/tradearea/GeoUtils.java:
```diff
@@ -0,0 +1,43 @@
+    // radians/degrees conversion
+    static double degreesToRadians = Math.PI/180;
+    static double radiansToDegrees = 180/Math.PI;
+    // earth radius in km
```
Same here

tdesfont replied:
Ok

MayeulRousselet commented on java-processors/src/main/java/com/dataiku/dip/plugins/tradearea/GeoUtils.java:
```diff
@@ -0,0 +1,43 @@
+    // earth radius in km
+    static double earthRadius = 6378.137;
+
+    public static GeoPoint.Coords computeDestinationPoint(double latitude, double longitude, double bearing, double distance){
```
Nice comment but not in the expected java format. See https://www.oracle.com/fr/technical-resources/articles/java/javadoc-tool.html for the specs. Anyway, the comment should be moved out of the method and follow this format:
```suggestion
/**
        Given a starting point latitude and longitude, compute latitude and longitude of the unique final point defined as the
        projection of the starting point on a circle of radius equal to `distance` (km) and with an angle equal to `bearing` (degrees).
        A bearing angle of 0 will result to output point being at North of starting point
        A bearing angle of 90 will result to output point being at West of starting point ...
        @param latitude: Latitude angle of the starting point in degrees
        @param  longitude: Longitude angle of the starting point in degrees
        @param bearing: Bearing angle in degrees (0 is north, 90 is west ... ), angle can be any real number
        @param distance: Distance in kilometers
       @return Coordinates of the resulting point
         */
    public static GeoPoint.Coords computeDestinationPoint(double latitude, double longitude, double bearing, double distance){
```

tdesfont replied:
Done

MayeulRousselet commented on java-processors/src/main/java/com/dataiku/dip/plugins/tradearea/AreaGenerator.java:
```diff
@@ -0,0 +1,126 @@
+import com.dataiku.dip.shaker.types.GeoPoint;
+import com.dataiku.dip.utils.DKULogger;
+
+abstract class AreaGenerator {
```
```suggestion
/**
    An abstract class expected by the TradeAreaProcessor to generate the polygons
     */
abstract class AreaGenerator {
```

tdesfont replied:
Ok

MayeulRousselet commented on java-processors/src/main/java/com/dataiku/dip/plugins/tradearea/AreaGenerator.java:
```diff
@@ -0,0 +1,126 @@
+
+}
+
+class RectangleAreaGenerator extends AreaGenerator {
```
```suggestion
/**
    Generates a rectangular area centered on an input geopoint coordinates.
    The parameters width, height, radius are expected to be in kilometers.
    If using miles in the processor, those distances must be converted to miles before.
     */
class RectangleAreaGenerator extends AreaGenerator {
```

tdesfont replied:
ok

MayeulRousselet commented on java-processors/src/main/java/com/dataiku/dip/plugins/tradearea/AreaGenerator.java:
```diff
@@ -0,0 +1,126 @@
+        }
+    }
+
+    public String generateArea(GeoPoint.Coords center) {
```
```suggestion
/**
        Generate a rectangular trade area centered on an input geopoint coordonates.
        @param center : Coordinates of the center geopoint
        @return The rectangular trade area expressed as a WKT polygon (string)
            example: `POLYGON((long1 lat1,long2 lat2, ...))`
         */
    public String generateArea(GeoPoint.Coords center) {
```

tdesfont replied:
ok

MayeulRousselet commented on java-processors/src/main/java/com/dataiku/dip/plugins/tradearea/AreaGenerator.java:
```diff
@@ -0,0 +1,126 @@
+        StringBuilder str = new StringBuilder();
+        str.append("POLYGON((");
+        // Computation of the diagonal angle of the rectangle (Must be in degree and Math.atan is radian)
+        double diagonalAngle = Math.atan(this.height/this.width)*180/Math.PI;
```
As this variable do not depends on the method parameter, declared it as a field and iinit it in the constructor.

tdesfont replied:
Ok

MayeulRousselet commented on java-processors/src/main/java/com/dataiku/dip/plugins/tradearea/AreaGenerator.java:
```diff
@@ -0,0 +1,126 @@
+        // Compute the latitude longitude of the four corners and fill the polygon String
+        for (int i = 0; i < angles.length; i++){
+            double angle;
+            angle = angles[i];
```
```suggestion
            double angle = angles[i];
```

tdesfont replied:
Ok

MayeulRousselet commented on java-processors/src/main/java/com/dataiku/dip/plugins/tradearea/AreaGenerator.java:
```diff
@@ -0,0 +1,126 @@
+            double angle;
+            angle = angles[i];
+            GeoPoint.Coords tmpCoords;
+            tmpCoords = GeoUtils.computeDestinationPoint(center.latitude, center.longitude, angle, this.radius);
```
```suggestion
            GeoPoint.Coords tmpCoords = GeoUtils.computeDestinationPoint(center.latitude, center.longitude, angle, this.radius);
```

tdesfont replied:
Ok

MayeulRousselet commented on java-processors/src/main/java/com/dataiku/dip/plugins/tradearea/AreaGenerator.java:
```diff
@@ -0,0 +1,126 @@
+            angle = angles[i];
+            GeoPoint.Coords tmpCoords;
+            tmpCoords = GeoUtils.computeDestinationPoint(center.latitude, center.longitude, angle, this.radius);
+            str.append(tmpCoords.longitude + " " + tmpCoords.latitude + ",");
```
Avoid temporary string creation by using multiple append.
```suggestion
str.append(tmpCoords.longitude);
str.append(" ");
str.append(tmpCoords.latitude);
str.append(",");
```

tdesfont replied:
Ok now

MayeulRousselet commented on java-processors/src/main/java/com/dataiku/dip/plugins/tradearea/AreaGenerator.java:
```diff
@@ -0,0 +1,126 @@
+        double diagonalAngle = Math.atan(this.height/this.width)*180/Math.PI;
+        double[] angles = {90-diagonalAngle, 90+diagonalAngle, 270-diagonalAngle, 270+diagonalAngle};
+        // Compute the latitude longitude of the four corners and fill the polygon String
+        for (int i = 0; i < angles.length; i++){
```
Start at i=1 and handle the initCoord specifically

tdesfont replied:
Ok

MayeulRousselet commented on java-processors/src/main/java/com/dataiku/dip/plugins/tradearea/AreaGenerator.java:
```diff
@@ -0,0 +1,126 @@
+            }
+        }
+        // Close the polygon by adding initial geopoint
+        str.append(initCoords.longitude + " " + initCoords.latitude);
```
Remove the "+"

tdesfont replied:
Ok now

MayeulRousselet commented on java-processors/src/main/java/com/dataiku/dip/plugins/tradearea/AreaGenerator.java:
```diff
@@ -0,0 +1,126 @@
+}
+
+class CircleAreaGenerator extends AreaGenerator {
+    /*
```
Use a class comment

tdesfont replied:
Ok now

MayeulRousselet commented on java-processors/src/main/java/com/dataiku/dip/plugins/tradearea/AreaGenerator.java:
```diff
@@ -0,0 +1,126 @@
+    }
+
+    public String generateArea(GeoPoint.Coords center) {
+        /*
```
Use the good method foramt

tdesfont replied:
Ok now

MayeulRousselet commented on java-processors/src/main/java/com/dataiku/dip/plugins/tradearea/AreaGenerator.java:
```diff
@@ -0,0 +1,126 @@
+            example: `POLYGON((long1 lat1,long2 lat2, ...))`
+         */
+        if (this.radius<=0){
+            logger.info("Detected invalid radius as input. Radius parameter should be greater than zero.");
```
Log the radius value

tdesfont replied:
Ok

MayeulRousselet commented on java-processors/src/main/java/com/dataiku/dip/plugins/tradearea/AreaGenerator.java:
```diff
@@ -0,0 +1,126 @@
+            example: `POLYGON((long1 lat1,long2 lat2, ...))`
+         */
+        if (this.radius<=0){
+            logger.info("Detected invalid radius as input. Radius parameter should be greater than zero.");
```
Log the radius value

tdesfont replied:
Ok

MayeulRousselet commented on java-processors/src/main/java/com/dataiku/dip/plugins/tradearea/AreaGenerator.java:
```diff
@@ -0,0 +1,126 @@
+        StringBuilder str = new StringBuilder();
+        str.append("POLYGON((");
+        // Compute the points on circle, angle step is set to 30 degrees as there are 12 points
+        for (int i = 0; i < 12; i++){
```
Factorize with the method for the rectangle.

tdesfont replied:
Did you decide to do that at the end ?

tdesfont replied:
Done

MayeulRousselet commented on java-processors/src/main/java/com/dataiku/dip/plugins/tradearea/AreaGenerator.java:
```diff
@@ -0,0 +1,126 @@
+        StringBuilder str = new StringBuilder();
+        str.append("POLYGON((");
+        // Compute the points on circle, angle step is set to 30 degrees as there are 12 points
+        for (int i = 0; i < 12; i++){
```
Create a constant for 12:
```suggestion
class CircleAreaGenerator extends AreaGenerator {
       static final int NB_OF_EDGES = 360;
       ....
       for (int i = 0; i < NB_OF_EDGES; i++){
```

tdesfont replied:
Ok

MayeulRousselet commented on java-processors/src/main/java/com/dataiku/dip/plugins/tradearea/GeoUtils.java:
```diff
@@ -0,0 +1,43 @@
+         */
+        latitude *= degreesToRadians;
+        longitude *= degreesToRadians;
+        bearing *= degreesToRadians;
```
Use radians for bearing and not degrees

tdesfont replied:
It might be a bit strange to have latitude, longitude in degrees and bearing in radians

MayeulRousselet replied:
Not necessarily if the variable name reflects it

tdesfont committed:
Potential fix for the null point exception
TradeAreaProcessor.java: Check nullity instead of length == 0

MayeulRousselet commented on java-processors/src/main/java/com/dataiku/dip/plugins/tradearea/TradeAreaProcessor.java:
```diff
@@ -0,0 +1,199 @@
+        private static final long serialVersionUID = -1;
+        public String inputColumn;
+        public String outputColumn;
+        // Custom parameters declaration of the processor
```
```suggestion
        /** 
          * Custom parameters declaration of the processor
          */
```

tdesfont replied:
Done

MayeulRousselet commented on java-processors/src/main/java/com/dataiku/dip/plugins/tradearea/TradeAreaProcessor.java:
```diff
@@ -0,0 +1,199 @@
+
+        @Override
+        public void validate() throws IllegalArgumentException {
+            TradeAreaProcessor.newGenerator(unitMode, shapeMode, radius, height, width);
```
It will be better to implement a static validate method on TradeAreaProcessor

tdesfont replied:
Not sure to understand, will commit the following
```
    static void validate(UnitMode unitMode, ShapeMode shapeMode, double radius, double height, double width){
        AreaGenerator areaGenerator = newGenerator(unitMode, shapeMode, radius, height, width);
    }
```

MayeulRousselet commented on java-processors/src/main/java/com/dataiku/dip/plugins/tradearea/TradeAreaProcessor.java:
```diff
@@ -0,0 +1,199 @@
+
+    @VisibleForTesting
+    static AreaGenerator newGenerator(UnitMode unitMode, ShapeMode shapeMode, double radius, double height, double width) {
+        logger.info("Instantiated new generator with argument: width="+width+" height="+height);
```
Better to use infoV to log variable arguments + log the radius
```suggestion
        logger.infoV("Instantiated new generator with argument: width=%f height=%f", width, height);
```

tdesfont replied:
Done

MayeulRousselet commented on java-processors/src/main/java/com/dataiku/dip/plugins/tradearea/TradeAreaProcessor.java:
```diff
@@ -0,0 +1,199 @@
+    static AreaGenerator newGenerator(UnitMode unitMode, ShapeMode shapeMode, double radius, double height, double width) {
+        logger.info("Instantiated new generator with argument: width="+width+" height="+height);
+        // Conversion factor from miles to km to enforce km as input
+        double milesToKm = 1.6093444978925633;
```
declares it as a class constant 
```suggestion
      static final double MILES_TO_KM = 1.6093444978925633;
```

tdesfont replied:
Done

MayeulRousselet commented on java-processors/src/main/java/com/dataiku/dip/plugins/tradearea/TradeAreaProcessor.java:
```diff
@@ -0,0 +1,199 @@
+        double milesToKm = 1.6093444978925633;
+        // Expected input distances in the area generators are in km
+        // If in miles, convert to km
+        logger.info("Using miles as standard unit for distances");
```
I think we use km as standard instead of miles

tdesfont replied:
Done

MayeulRousselet commented on java-processors/src/test/java/com/dataiku/dip/plugins/tradearea/CircleAreaGeneratorTest.java:
```diff
@@ -0,0 +1,18 @@
+
+import com.dataiku.dip.shaker.types.GeoPoint;
+
+class CircleAreaGeneratorTest {
```
Add tests for invalid input etc, as you did for the Rectangle

tdesfont replied:
Done

MayeulRousselet commented on java-processors/src/test/java/com/dataiku/dip/plugins/tradearea/RectangleAreaGeneratorTest.java:
```diff
@@ -0,0 +1,82 @@
+        String actualResult = null;
+        RectangleAreaGenerator generator = new RectangleAreaGenerator(width, height);
+        String str = "Bad input";
+        GeoPoint.Coords coords = GeoPoint.convert(str);
```
GeoPoint.convert will return null for "Bad Input", therefore you don't test your code as you never enter in the if.
better to test generator.generateArea with a null input.

tdesfont replied:
Right ! Fixed

tdesfont committed:
Set diagonalAngle as a class variable to avoid unnecessary computation. Used chained addition to String builder. Enhanced explanation of the trade area processor.

MayeulRousselet commented on java-processors/src/test/java/com/dataiku/dip/plugins/tradearea/RectangleAreaGeneratorTest.java:
```diff
@@ -0,0 +1,82 @@
+        String actualResult = null;
+        RectangleAreaGenerator generator = new RectangleAreaGenerator(width, height);
+        String str = "POINT(-73.9723, 40.64749)";
+        GeoPoint.Coords coords = GeoPoint.convert(str);
```
Here you only test for GeoCoord.convert, which must be tested beforehand

tdesfont replied:
Should I remove any tests that test GeoCoord.convert ?

tdesfont replied:
Done in 4fe5ecb

MayeulRousselet commented on java-processors/src/test/java/com/dataiku/dip/plugins/tradearea/RectangleAreaGeneratorTest.java:
```diff
@@ -0,0 +1,82 @@
+        RectangleAreaGenerator generator = new RectangleAreaGenerator(width, height);
+        String str = "POINT(-73.9723 1000)";
+        GeoPoint.Coords coords = GeoPoint.convert(str);
+        System.out.println(coords);
```
Remove the System.out

tdesfont replied:
Done

MayeulRousselet commented on java-processors/src/test/java/com/dataiku/dip/plugins/tradearea/RectangleAreaGeneratorTest.java:
```diff
@@ -0,0 +1,82 @@
+        String str = "POINT(-73.9723 1000)";
+        GeoPoint.Coords coords = GeoPoint.convert(str);
+        System.out.println(coords);
+        if (coords != null){
```
Coords il never null, remove the test

tdesfont replied:
Done

MayeulRousselet commented on java-processors/src/test/java/com/dataiku/dip/plugins/tradearea/RectangleAreaGeneratorTest.java:
```diff
@@ -0,0 +1,82 @@
+        String str = "POINT(-73.9723 40.64749)";
+        GeoPoint.Coords coords = GeoPoint.convert(str);
+        System.out.println(coords);
+        if (coords != null){
```
Coords will not be null, remove the if and remove the System.out.println

tdesfont replied:
Done

tdesfont committed:
Update java-processors/src/main/java/com/dataiku/dip/plugins/tradearea/TradeAreaProcessor.java

Co-authored-by: MayeulRousselet <54799490+MayeulRousselet@users.noreply.github.com>

tdesfont committed:
Fix class comments syntax, logging.

tdesfont committed:
Merge branch 'feature/gradle-trade-area-processor' of github.com:dataiku/dss-plugin-geospatial-toolkit into feature/gradle-trade-area-processor

tdesfont committed:
Refactor the loop for angles and edges creation

tdesfont committed:
Convert bearing angle to radians

tdesfont committed:
Switched to logger.infoV

tdesfont committed:
Extracted MILES_TO_KM and fix the logging for unit of distances

tdesfont committed:
Created unit tests for CircularAreaGenerator

tdesfont committed:
Handle null and adapt tests

tdesfont committed:
Enhance tests

tdesfont committed:
Refactoring of the wkt string building loop. Created a class PolygonBuilder.

MayeulRousselet commented on java-processors/src/main/java/com/dataiku/dip/plugins/tradearea/TradeAreaProcessor.java:
```diff
@@ -177,25 +184,34 @@ public void init() throws Exception {
     }
+
+    static void validate(UnitMode unitMode, ShapeMode shapeMode, double radius, double height, double width){
+        AreaGenerator areaGenerator = newGenerator(unitMode, shapeMode, radius, height, width);
```
You should have a specific method to validate without creating a new AreaGenerator.
You should check for shapeMode, unitMode, radius > 0 (if cricle) and height & witdh (if rectancle).

tdesfont replied:
Done in 371c918

MayeulRousselet commented on java-processors/src/main/java/com/dataiku/dip/plugins/tradearea/AreaGenerator.java:
```diff
@@ -98,28 +99,26 @@ public CircleAreaGenerator(double radius){
+            return null;
+        }
 
         if (this.radius<=0){
```
No need to check the radius if it was checked before in the validation, the value can't be changed.

tdesfont replied:
done in e0f3eaf

MayeulRousselet commented on java-processors/src/main/java/com/dataiku/dip/plugins/tradearea/AreaGenerator.java:
```diff
@@ -45,35 +46,35 @@ public RectangleAreaGenerator(double width, double height){
+        if (center == null){
+            return null;
+        }
         if (this.radius<=0 || this.width<=0 || this.height<=0){
```
No need to check for these value as it should be done by the validation.

tdesfont replied:
done in e0f3eaf

MayeulRousselet commented on java-processors/src/main/java/com/dataiku/dip/plugins/tradearea/PolygonBuilder.java:
```diff
@@ -0,0 +1,26 @@
+
+public class PolygonBuilder {
+
+    StringBuilder wktPolygon;
```
add final (the variable won't be reassigned after the constructor has been called)
```suggestion
    final StringBuilder wktPolygon;
```

tdesfont replied:
Done in 4a72e51

MayeulRousselet commented on java-processors/src/main/java/com/dataiku/dip/plugins/tradearea/PolygonBuilder.java:
```diff
@@ -0,0 +1,26 @@
+public class PolygonBuilder {
+
+    StringBuilder wktPolygon;
+    GeoPoint.Coords firstPoint;
```
Add final

tdesfont replied:
Done in f479de8

MayeulRousselet commented on java-processors/src/main/java/com/dataiku/dip/plugins/tradearea/PolygonBuilder.java:
```diff
@@ -0,0 +1,26 @@
+
+import com.dataiku.dip.shaker.types.GeoPoint;
+
+public class PolygonBuilder {
```
If there is no call to lineTo, the resulting polygon will be bogus (the 2 same points without a comma between the two).
Either document in the class comment that a lineTo call is mandatory, or handle the case.

tdesfont replied:
Done in 72e5797

tdesfont committed:
Use specific function to check input parameters

tdesfont committed:
Remove unused check on parameters

tdesfont committed:
Set wktPolygon variable to final

tdesfont committed:
Set wktPolygon variable to final

tdesfont committed:
Document PolygonBuilder class

tdesfont committed:
Add details to user message for invalid distance parameters

MayeulRousselet commented on java-processors/src/main/java/com/dataiku/dip/plugins/tradearea/AreaGenerator.java:
```diff
@@ -0,0 +1,146 @@
+                logger.info("Rectangle Area Generator: Received invalid parameters as input.");
+                logger.infoV("Got width= {}", width);
+                logger.infoV("Got height= {}", height);
+                throw new IllegalArgumentException(
```
I am not sure about the throw as it will result in a ugly message on the front: See for example:
<img width="1386" alt="Capture d’écran 2021-02-08 à 9 51 35 PM" src="https://user-images.githubusercontent.com/54799490/107280038-1b24e080-6a58-11eb-872d-c1b60c197d5d.png">

I've dig in the DIP code to look for validation in processor and very few of them do it.
I think we should only check for shapeMode and UnitMode and raise an exception if they are not known, and silently do nothing if radius or width or height are <=0.

tdesfont replied:
See 7a41321
Added ignore on invalid radius or width/height value, removed the throw error message for invalid width/height/radius

MayeulRousselet commented on java-processors/src/main/java/com/dataiku/dip/plugins/tradearea/TradeAreaProcessor.java:
```diff
@@ -0,0 +1,217 @@
+
+        @Override
+        public void validate() throws IllegalArgumentException {
+            TradeAreaProcessor.validate(unitMode, shapeMode, radius, height, width);
```
Directly call AreaGenerator. checkInputParams here and remove TradeAreaProcessor.validate

tdesfont replied:
Done in b225e00

MayeulRousselet commented on java-processors/src/main/java/com/dataiku/dip/plugins/tradearea/AreaGenerator.java:
```diff
@@ -0,0 +1,146 @@
+ */
+class RectangleAreaGenerator extends AreaGenerator {
+
+    double width;
```
declare each field as final

tdesfont replied:
Done in 8a82db6

MayeulRousselet commented on java-processors/src/main/java/com/dataiku/dip/plugins/tradearea/AreaGenerator.java:
```diff
@@ -0,0 +1,146 @@
+        }
+
+        // Declaration of the final result
+        StringBuilder str = new StringBuilder();
```
str is not used anymore, remove it

tdesfont replied:
Done in 8214dc5

MayeulRousselet commented on java-processors/src/main/java/com/dataiku/dip/plugins/tradearea/AreaGenerator.java:
```diff
@@ -0,0 +1,146 @@
+    static final int NB_OF_EDGES = 12;
+    static final double ANGLE_STEP = 2*Math.PI/NB_OF_EDGES;
+
+    double radius;
```
declare final

tdesfont replied:
Done

MayeulRousselet commented on java-processors/src/main/java/com/dataiku/dip/plugins/tradearea/AreaGenerator.java:
```diff
@@ -0,0 +1,146 @@
+     * @return The rectangular trade area expressed as a WKT polygon (string)
+     */
+    public String generateArea(GeoPoint.Coords center) {
+        if (center == null){
```
```suggestion
        if (center == null || diagonalAngle==0){
```

tdesfont replied:
Done in 3b270d4

MayeulRousselet commented on java-processors/src/main/java/com/dataiku/dip/plugins/tradearea/AreaGenerator.java:
```diff
@@ -0,0 +1,146 @@
+     * @return The circular trade area expressed as a WKT POLYGON (String `POLYGON((long1 lat1,long2 lat2, ...))`)
+     */
+    public String generateArea(GeoPoint.Coords center) {
+        if (center == null){
```
```suggestion
        if (center == null || radius==0){
```

tdesfont replied:
Done in e30115b

MayeulRousselet commented on java-processors/src/main/java/com/dataiku/dip/plugins/tradearea/AreaGenerator.java:
```diff
@@ -0,0 +1,146 @@
+        if (width <= 0 || height <= 0){
+            logger.info("Rectangle Area Generator: Received invalid parameters as input.");
+            logger.infoV("Got width= {}", width);
+            logger.infoV("Got height= {}", height);
```
init this.height, this.width and this.diagonalAngle to 0

tdesfont replied:
Done in a8002c2

MayeulRousselet commented on java-processors/src/main/java/com/dataiku/dip/plugins/tradearea/AreaGenerator.java:
```diff
@@ -0,0 +1,146 @@
+    double radius;
+
+    public CircleAreaGenerator(double radius){
+        this.radius = radius;
```
```suggestion
        this.radius = Math.max(0, radius);
```

tdesfont replied:
Done in 16c0519

MayeulRousselet commented on java-processors/src/main/java/com/dataiku/dip/plugins/tradearea/AreaGenerator.java:
```diff
@@ -0,0 +1,146 @@
+            throw new IllegalArgumentException("Invalid processing mode: " + shapeMode);
+        }
+    };
+
```
Check for UnitMode too

tdesfont replied:
Done

MayeulRousselet commented on java-processors/src/main/java/com/dataiku/dip/plugins/tradearea/TradeAreaProcessor.java:
```diff
@@ -0,0 +1,217 @@
+            }
+    }
+
+    static void validate(UnitMode unitMode, ShapeMode shapeMode, double radius, double height, double width){
```
TO remove

tdesfont replied:
Done

MayeulRousselet commented on java-processors/src/test/java/com/dataiku/dip/plugins/tradearea/CircleAreaGeneratorTest.java:
```diff
@@ -0,0 +1,54 @@
+    }
+
+    @Test
+    public void testInvalidFormat() {
```
You don't have to test for that, It should be tested in DSS.

tdesfont replied:
Done

MayeulRousselet commented on java-processors/src/test/java/com/dataiku/dip/plugins/tradearea/CircleAreaGeneratorTest.java:
```diff
@@ -0,0 +1,54 @@
+        String str = "POINT(-73.9723 1000)";
+        GeoPoint.Coords coords = GeoPoint.convert(str);
+        String actualResult = generator.generateArea(coords);
+        assertNotNull(actualResult);
```
You should test against a real expected value.

tdesfont replied:
Done in 2fe6e21

MayeulRousselet commented on java-processors/src/test/java/com/dataiku/dip/plugins/tradearea/RectangleAreaGeneratorTest.java:
```diff
@@ -0,0 +1,56 @@
+    }
+
+    @Test
+    public void testInvalidRange() {
```
You don't test an invalid range here, you do the same test as in testStandardGenerator. Delete this test

tdesfont replied:
Done in f9fef19

MayeulRousselet commented on java-processors/src/main/java/com/dataiku/dip/plugins/tradearea/TradeAreaProcessor.java:
```diff
@@ -0,0 +1,217 @@
+        if (!StringUtils.isBlank(params.outputColumn)) {
+            outputColumn = getCf().column(params.outputColumn, ProcessorRole.OUTPUT_COLUMN);
+        } else {
+            outputColumn = cd;
```
There is a bug here.
Let say I have three columns in my dataset (lon, lat and location).
I choose lon as an input col, and no output column, then the value of the lon columns are wiped out.
Cf screenshot below:
<img width="1387" alt="Capture d’écran 2021-02-08 à 10 43 50 PM" src="https://user-images.githubusercontent.com/54799490/107285239-262f3f00-6a5f-11eb-8a85-98f4e7772b9f.png">

tdesfont replied:
Can't reproduce the bug.

tdesfont replied:
This is the standard behaviour of native processor, in our case, if null values is returned and there is a value in the input column, we decide to leave the input column value in the result column.

<img width="470" alt="Screenshot 2021-02-15 at 11 25 18" src="https://user-images.githubusercontent.com/17903135/107934540-7d15a680-6f80-11eb-9364-173e68be5a72.png">

tdesfont replied:
Fixed in 1c592f4.

tdesfont committed:
CircleAreaGeneratorTest: Removed tests on the GeoSpatial Coordinates parsing
RectangleAreaGeneratorTest: Removed failing test on the input value equal to zero (will be dealt with later)

tdesfont committed:
AreaGenerator:
- Added check on unit mode in addition to shape mode in the input checker.
- Added silently return of null values if input parameters are less or equal than zero.

tdesfont committed:
TradeAreaProcessor.java: Removed TradeAreaProcessor.java intermediate check functions to directly call the AreaGenerator.checkInputParams

tdesfont committed:
AreaGenerator: Set class variable to final as they have fixed value once initialized

tdesfont committed:
AreaGenerator: Removed call to StringBuilder since it is declared and used in the new PolygonBuilder class

tdesfont committed:
AreaGenerator: Added check on diagonalAngle

tdesfont committed:
AreaGenerator: Refactored generateArea input checks

tdesfont committed:
AreaGenerator: Added variables initialisation to zero, removed final declaration

tdesfont committed:
AreaGenerator: Added instantiation of the radius value

tdesfont committed:
CircleAreaGeneratorTest: Added expected result to compare to an actual value
RectangleAreaGeneratorTest: Same

tdesfont committed:
Removed testInvalidRange in both classes

tdesfont committed:
CircleAreaGeneratorTest/RectangleAreaGeneratorTest: Set the test on null value again

tdesfont committed:
Fix bug on column overwrite for null values.

clemencebic commented:
Here are a few things we could improve on the processor : 
- we already said we want to increase the number of edges for the circle (we are still using 12 edges)
- we should reverse the order of the "distance unit" and "shape" parameters (the distance unit parameter should be closer to radius en width parameter)
- for radius, width and height I would add the following indication : "radius (for circle shape)", 'width (for rectangle shape)' ...
- Can we prevent the user from putting negative values in the width, height and radius parameters ?  Or at least raise an error
- Is it possible to add warnings ? For example if shape=circle and radius=0 could we put a warning "A 0 radius will generate empty values)
- In the processor description, we should also mention the "radius", "width" and 'height' parameters
- UI and wording of the trade area : 
![image](https://user-images.githubusercontent.com/18233438/108042902-b7df1380-7040-11eb-9ce3-2c080bef6002.png)
1. We should keep the same type of syntax : "Generate" instead of "Generating"
2. We should keep the same UI than the other processors, that is "Generate a trade area" in bold, and "merchant location" in red in my example

tdesfont committed:
[ch59576] Increase number of vertices for circular trade area zone
AreaGenerator.java: Set constant number of edges to 120

tdesfont committed:
Invert unit mode and shape mode input params in the UI to keep distance unit close to radius

tdesfont committed:
Add hint in radius/height/width description, describe the type of Shape precisely

tdesfont committed:
Wording for processor description

tdesfont committed:
Wording, add precision in the step description about radius, width, height

tdesfont commented:
- [x] Increase the number of edges for the circle (Set to 36)
- [x] Reverse the order of the "distance unit" and "shape" parameters
- [x] add the following indication : "radius (for circle shape)", 'width (for rectangle shape)' ...
- [x] Prevent the user from putting negative values in the width, height and radius parameters 
- [ ] Add warnings ? For example if shape=circle and radius=0 could we put a warning "A 0 radius will generate empty values)
- [x] In the processor description, we should also mention the "radius", "width" and 'height' parameters
- [x] UI and wording of the trade area: Keep the same type of syntax : "Generate" instead of "Generating"
- [ ]  UI and wording of the trade area: We should keep the same UI than the other processors, that is "Generate a trade area" in bold, and "merchant location" in red in my example

tdesfont committed:
Add error on negative distance values encountered as input params

tdesfont committed:
AreaGenerator.java: Fixed wording of error warning

tdesfont committed:
processor.js: Add default message if no geopoint column is selected

tdesfont committed:
Merge branch 'master' into feature/gradle-trade-area-processor
